### DISCUSSION TITLE
Welcome!

### DISCUSSION THREAD

**Discussion by anjohnson** [epics-base/epics-base#239] [Codeathon 2022]
Welcome to the Core Track of the hybrid [2022 EPICS Codeathon](https://epics.anl.gov/meetings/codeathon-22/index.php), hosted in-person at SLAC for those able to attend in-person, or online for those who can't. This GitHub Discussion category in the EPICS Base project is for questions and feedback of any nature related to the Codeathon, the other tracks should have their own discussion area for more focussed technical questions, but feel free to use this one if you have a more general question.

You can find a list of the Core C/C++ projects on [this Wiki page](https://github.com/epics-base/epics-base/wiki/Codeathon-project-ideas). Each project includes a description, the name of the person who posted it, links to other information about it, and an indication on how you can "claim" the project before you start working on it. If it just says 'Claimed by: ?' that means edit the wiki page and replace the ? with your name or GitHub handle, but there are other ways for some projects, so please check the specific instructions so we don't have two or more solutions being developed to the same task.

In most cases you should be creating a Pull Request (PR) with your solution. If you want to ask for feedback before you've finished the work, create the PR and mark it as Draft, which will tell your reviewers not to spend time telling you what's missing (yet). You can continue to push commits to the same branch on your GitHub fork and they will be incorporated into your PR automatically. Once you think you've completed the task (and added entries to the Release Notes and any other documentation that's appropriate) you can adjust the PR to remove the Draft status.

In addition to developing code, please try to review at least one other Pull Request for every PR you create (more if you enjoy that, studying other people's code changes can be good learning). The expression "many eyes make all bugs shallow" at least shows the right sentiment, and the Core Developer group members appreciate not having to do all the code reviews for this week.

Thank-you for taking part and spending your time helping to make EPICS better!

- Andrew Johnson, for the EPICS Core Developers Group

**Comment by minijackson** [comment]
Hello, and thanks for the organization!

Me and my colleague (@stephane-plus, who is still battling with GitHub because his account was flagged within 10mins of signing up), took the liberty of taking 3 subjects from the Wiki page (the iocsh command help, the CI cross-compilation, and the base regression testing using containers). I assume that when unspecified, the task is "up for grabs" by anyone, is that correct? If not, we don't mind changing back the page.

Thanks again!

**Comment by simon-ess** [comment]
Hello all -

We will be hosting a parallel session on European time (CEST), 9 hours earlier than the main event at SLAC. We will be following the same schedule, and using the same task lists linked above.

The session will be at least partially in person on site at ESS, and held over zoom as well. Please contact @dirk-zimoch for a link to the zoom meeting if you would like to join online.

Cheers,

**Reply by dirk-zimoch** [reply]
We Europeans now use the "normal" Zoom link. It works during our day time, too. We meet daily 9:00 European time and if possible at 18:00 (= 9:00 AM PDT ).

**Comment by stephane-cea** [comment]
Hi everyone!

About the [EPICS container images](https://github.com/epics-base/epics-base/wiki/Codeathon-project-ideas#create-set-of-container-images-with-different-epics-base-releases-for-testing) topic:

@ralphlange made me a few suggestions, I think it would be nice to share them here, for anyone to see and eventually comment:
- We need to define a few use cases in order to get an idea of the general direction to take. These images should be easy to build upon, so that we can e.g. add a modules-container on top of an EPICS-base-container (R7.0 or R3.15).
- We need to agree on locations of modules, users, the file layout of EPICS stuff in the images.
- For unit testing and automation, it is probably most appropriate to mount the working directory with the software under test into the container, then build the software against the containerized base and run the tests.
- For cross-compatibility testing of Channel Access, we would need one container that runs some standard "testing IOC" with a well-known database and another one that has the standard client tools available.

@ralphlange also pointed me to these relevant/similar projects:
- https://github.com/pklaus/docker-epics-directory (from Philipp Klaus, at FAIR, also see https://epics.anl.gov/tech-talk/2020/msg01229.php)
- https://indico.lightsource.ca/event/2/timetable/#49-kubernetes-for-epics-iocs (from Giles Knap, at Diamond)
https://github.com/prjemian/epics-docker
- (I also found this one: https://github.com/prjemian/epics-docker from Pete R Jemian, at APS)

Also, I was told that @mdavidsaver probably has the best idea of what is needed for testing and has the most experience running tests in CI context. @mdavidsaver: any comments about these suggestions?

This morning I managed to create a dockerfile ready to build a R7.0 EPICS-base-container, another one to build a R3.15 EPICS-base-container, and a last one to build a synApps-container (based on either R7.0 or R3.15). Is there a dedicated GitHub repo where I could share those dockerfiles? I can create one if needed. Also, is there a dedicated [hub.docker](hub.docker.com) space where I could host the resulting Docker images ? I can also create one if needed.

**Reply by fuse-energy-dev** [reply]
I replied to the announcements email, but the [message](https://github.com/epics-base/epics-base/discussions/239#discussioncomment-2716881)  says it all: I have multi-arch images for arm64 and amd64/x86_64 
based on 3.15.6, as well as synapps6 images and some that have specific modules: calc, asyn, sncseq, and pydevice. I can create a public docker hub repository for reference if anyone is interested. Just holler.

**Reply by mdavidsaver** [reply]
> * We need to define a few use cases ...

I think there are two parts.  First is simply to run the unittests as part of the build process.  The second is to facilitate interoperability/compatibility testing.  So that eg. I could ~quickly start a 3.14.12 IOC to test against some changes I'm making to 7.0.

So I'm not thinking about automating cross-compatibility testing.  At least not at this point.

>  ... Is there a dedicated GitHub repo ... is there a dedicated [hub.docker](hub.docker.com) space ...

No.  This is a new area.  Such could be created.  Though, I would suggest prototyping under individual accounts.

**Reply by stephane-cea** [reply]
Here are the results:
- https://github.com/stephane-cea/epics-containers
- https://hub.docker.com/repository/registry-1.docker.io/stephanecea/epics/tags?page=1&ordering=last_updated

As suggested by @mdavidsaver:
- I will create some centos7 based containers (if you want other distros, then please ask)
- I will also produce both amd64 and arm64 containers
- I will try to find a way to run arm containers on amd host

@fuse-energy-dev: if you can share your dockerfiles, it would be great!

If anyone has more suggestion, then please share!

**Reply by stephane-cea** [reply]
Some updates about the EPICS container images topic:
- I added some CentOS 7 based containers
- I reduced the size of the containers (between 100 MB and 500 MB, depending on the containers, instead of >1 GB)
- I have managed to build and run linux/arm64 containers on linux/amd64 hosts (see [the documentation](https://github.com/stephane-cea/epics-containers/blob/master/README.md) for more details)

All those containers take quite some time to build, especially the arm64 ones (which are emulated on my amd64 host, causing slow compilation times). At the moment I haven't re-build them all. So before the end of the day (and the and of the codeathon event) I will try to setup a CI to build them for me.

**Comment by gauge2** [comment]
I'm not participating in the codeathon, but I actually have created
multi-arch docker images (linux/amd64 and linux/arm64) from 3.15.6 and gone
on to build support images for SynApps6, and for custom module support.
If there is interest in looking at them, I can make a public hub.docker.com
repository to have a look. I use them to build on top of. Let me know what
you think.



*Josh Fiddler*

Senior Controls and Data Developer

+1-450-245-1065

www.f.energy

Linkedin <https://www.linkedin.com/in/josh-fiddler/>

**Comment by marciodo** [comment]
Hello, all. @anjohnson, your introduction text mentions "Each project includes a description, the name of the person who posted it, links to other information about it, and an indication on how you can "claim" the project before you start working on it". I'd like to claim "Improve the CAS error message generated when an IOC gets port-scanned", but I see no description about how to claim it.

**Reply by anjohnson** [reply]
Hi Marcio, if it just has 'Claimed by: ?' that means edit the wiki page and replace the `?` with your name or GitHub handle — sorry if that wasn't very clear (I'll add that to the intro above).

The [CAS error message project](https://github.com/epics-base/epics-base/wiki/Codeathon-project-ideas#improve-the-cas-error-message-generated-when-an-ioc-gets-port-scanned) has a [PR from last year](https://github.com/epics-base/epics-base/pull/141) which generated some discussion and changes, but unfortunately we never resolved them well enough to merge the result. I would like to refer you to @mdavidsaver to decide how best to move forward with this, I suggested the project but he's the technical expert in the IOC's networking code.

**Comment by anjohnson** [comment]
Daniel Flath wants to hold a close-out session on Thursday morning, with a summary of what we have achieved. In previous events we've had a session where everyone discusses their own work, but that might not work for the EU participants. Would someone from that session be willing to join our Zoom meeting and talk about what you guys have been doing? I'm thinking 9am SLAC time but we could make it up to an hour earlier if necessary. Please discuss amongst yourselves and respond here.

For the on-site participants, please talk to me about availability at 8am, 8:30 or 9am, and make sure I'm aware of everything you've been doing if it isn't documented as a GitHub Pull Request or by an entry here. You don't have to have finished it by then though!

**Reply by echandler-anl** [reply]
Is there going to be a close-out session today at 9am SLAC time?

**Comment by tboegi** [comment]
(Moving my wrongly comment here)
I would like to continue with the leftover from last year:
"Accessing Record fields from asyn driver"
https://epics.anl.gov/tech-talk/2021/msg00404.php

**Reply by tboegi** [reply]
Some progress made here:
The driver can put meta data into the asyn param-library.
Fiddling these values into the record fields does not work (yet)

**Comment by tboegi** [comment]
Another thing:
IPv6 for EPICS base.
@dirk-zimoch and @tboegi have done some effort to get "IPv6 for channel access" compiling
under RTEMS, vxworks and Windows.
https://github.com/tboegi/epics-base/branches
However, Timo pointed out, that we have pvxs, that is IPv6 prepared:
https://mdavidsaver.github.io/pvxs/releasenotes.html#
In that light, how much effort should be spend on ca/IPv6 ?
Some cleanup is needed anyway.
Looking at pvxs: there are good ideas how to improve compatiblity,
since today neither pvxs nor pcas/cacm are compiling against the IPv6 patches.

**Reply by tboegi** [reply]
Some fixes are floating around here in local branches; will be pushed soonish.
We may call the current branch a working prototype.
In the long run, a lot of changes need to be re-fined.

**Comment by hjunkes** [comment]
Now I'm finally in on it too. "Better late than never ;-)". 
I will now set about bringing RTEMS6 with EPICS7 to the MVME3100. Clean up a little the things I've done so far (RTEMS5 with MVME3100 and RTEMS6 with libbsd and nfs4, etc.) and try to integrate Miroslaw Dach's extensions (readline, etc.).

**Comment by simon-ess** [comment]
I have created a pull request for the compress record: https://github.com/epics-base/epics-base/pull/252

**Comment by echandler-anl** [comment]
I'd like to work on allowing PV requests with multiple fields from from the same sub-structure.  Right now the behavior is like this:

```
> pvget -r 'obj.x,obj.y' TESTPV
failed to parse request string: Can't construct Structure, duplicate fieldName obj
```

**Reply by echandler-anl** [reply]
Nevermind, it looks like this is by design.  The correct syntax is 'obj{x,y}'

**Comment by bhill-slac** [comment]
@ralphlange and others interested in CA gateway issues, please see these ca-gateway PRs: 
https://github.com/epics-extensions/ca-gateway/pull/41
https://github.com/epics-extensions/ca-gateway/pull/42
https://github.com/epics-extensions/ca-gateway/pull/43
https://github.com/epics-extensions/ca-gateway/pull/44
https://github.com/epics-extensions/ca-gateway/pull/45
